### PR TITLE
Fix e2e lerna publish (#3)

### PR DESCRIPTION
**Pull request by cgewecke**
(**DO NOT MERGE** until making suggested change in the comment below)

Unfortunately #3157 was only ~%25 correct :/ 

PR fixes several problems:
+ The packages were not being published by the lerna command correctly - they used `exec` (obviously wrong). I've had to upgrade Lerna to 3.x to get this to work.
+ Unexpectedly, truffle/contract mostly uses a [Web3 from a different truffle module][1] than itself, even though Web3 is a direct dependency there.  Both modules need to be modified to get the tests to consume the virtual publication correctly

I've added a 'proof' that this works - there's an incorrect console.log line in `web3-core-helpers` which is printed out while truffle's tests run. So we know their code is exercising the published copy.
It needs to be removed before merging..

Lastly, some of Truffle's tests are failing with an error that's already been reported at [truffle 2512][2]. I don't think any recent changes here are the cause because they [have not updated to 1.2.2 yet][3]. 

[1]: https://github.com/trufflesuite/truffle/blob/develop/packages/contract/lib/contract/constructorMethods.js#L219-L221
[2]: https://github.com/trufflesuite/truffle/issues/2512
[3]: https://github.com/trufflesuite/truffle/pull/2488

### CODE REVIEW

File: packages/web3-core-helpers/src/formatters.js [hunk 1]
```diff
@@ -219,6 +219,7 @@ var outputTransactionReceiptFormatter = function (receipt){
     if(typeof receipt.status !== 'undefined' && receipt.status !== null) {
         receipt.status = Boolean(parseInt(receipt.status));
     }
+    console.log('using new web3 at to format transaction receipt')
```
Comment (cgewecke) [inline packages/web3-core-helpers/src/formatters.js:222]: Edit this file to remove this line. It proves that the bundle is being published/used correctly [in travis here](https://travis-ci.com/cgewecke/web3.js/jobs/251293678#L1669-L1676).

**Review by nivida — CHANGES_REQUESTED** [review]
The truffle e2e container is failing during the bootstrapping process:

Bildschirmfoto 2019-10-31 um 12 31 41

**Comment by nivida** [comment]
@cgewecke The whole ``e2e`` tests concept should probably be documented somewhere here in the repository otherwise is it hard to follow up our current thoughts in some months (I will add a build pipeline section to the contributor's section of the readme in #3160 as well).

**Comment by cgewecke** [comment]
> The whole e2e tests concept should probably be documented somewhere here in the repository otherwise is it hard to follow up our current thoughts in some months (I will add a build pipeline section to the contributor's section of the readme in #3160 as well

I agree, this is a good idea.

File: scripts/e2e.npm.publish.sh [hunk 2]
```diff
@@ -34,11 +34,30 @@ npx npm-auth-to-token \
...
+if [ -z "$BRANCH" ]; then
 
-# `npm publish`
-npx lerna exec --concurrency 1 -- npm publish \
-  --tag e2e \
+  BRANCH=$TRAVIS_BRANCH
+
+fi
+
+git checkout $BRANCH --
```
Comment (cgewecke) [inline scripts/e2e.npm.publish.sh:45]: The build is failing here because 
+ Lerna [won't publish to the virtual registry][1] when the git head is in a detached state. 
+ It's necessary to explicitly checkout a branch to trick it
+ The branch isn't present when a fork's changes are proposed for the upstream

I'm going to resolve this temporarily by merging into a staging branch at web3 to make sure the CI runs as expected here. It would be nice if this ran for outside contributors, but perhaps is not critical that it does either because this kind of test is more of a pre-publication tests that's use internally anyway. 

https://github.com/lerna/lerna/issues/1595#issuecomment-422532097
